screensaver.turnoff/screensaver.py

```diff
@@ -1,164 +1,300 @@
+        func(self.display.get('function'), *self.display.get('args_on'))
+
+        # Clean up everything
+        self.exit()
+
+    def exit(self):
+        ''' Clean up function '''
+        if hasattr(self, 'monitor'):
+            del self.monitor
+
         self.close()
+#        del self
 
-if __name__ == '__main__':
-    addon = xbmcaddon.Addon()
-
-    addon_name = addon.getAddonInfo('name')
-    addon_path = addon.getAddonInfo('path')
-    addon_icon = addon.getAddonInfo('icon')
-    display_method = addon.getSetting('display_method')
-    power_method = addon.getSetting('power_method')
-    logoff = addon.getSetting('logoff')
-    mute = addon.getSetting('mute')
-
-    # Turn off display
-    if display_method != 0:
-        log_notice(msg='Turn display signal off using method %s' % display_method)
-    if display_method == '1':  # CEC (built-in)
-        run_builtin('CECStandby')
-    elif display_method == '2':  # No Signal on Raspberry Pi (using vcgencmd)
-        run_command(['vcgencmd', 'display_power', '0'])
-    elif display_method == '3':  # DPMS (built-in)
-        run_builtin('ToggleDPMS')
-    elif display_method == '4':  # DPMS (using xset)
-        run_command(['xset', 'dpms', 'force', 'off'])
-    elif display_method == '5':  # DPMS (using vbetool)
-        run_command(['vbetool', 'dpms', 'off'])
-    elif display_method == '6':  # DPMS (using xrandr)
-        # NOTE: This needs more outside testing
-        run_command(['xrandr', '--output CRT-0', 'off'])
-    elif display_method == '7':  # CEC on Android (kernel)
-        # NOTE: This needs more outside testing
-        run_command(['su', '-c', 'echo 0 >/sys/devices/virtual/graphics/fb0/cec'], shell=True)
-
-
-    # FIXME: Screensaver always seems to log off when logged in ?
-    # Log off user
-    if logoff == 'true':
-        run_builtin('System.Logoff()')
-
-    # Mute audio
-    if mute == 'true':
-        run_builtin('Mute')
 
+class TurnOffMonitor(Monitor, object):
+    ''' This is the monitor to exit TurnOffScreensaver '''
+
+    def __init__(self, **kwargs):  # pylint: disable=super-init-not-called
+        ''' Initialize monitor '''
+        self.action = kwargs.get('action')
+
+    def onScreensaverDeactivated(self):  # pylint: disable=invalid-name
+        ''' Perform cleanup function '''
+        self.action()
+
+
+ADDON = Addon()
+ADDON_NAME = to_unicode(ADDON.getAddonInfo('name'))
+ADDON_ID = to_unicode(ADDON.getAddonInfo('id'))
+ADDON_PATH = to_unicode(ADDON.getAddonInfo('path'))
+ADDON_ICON = to_unicode(ADDON.getAddonInfo('icon'))
+
+DEBUG_LOGGING = get_global_setting('debug.showloginfo')
+
+if __name__ == '__main__':
     # Do not start screensaver when command fails
-    screensaver = Screensaver('screensaver-turnoff.xml', addon_path, 'default')
-    screensaver.doModal()
-    del screensaver
+    TurnOffDialog('gui.xml', ADDON_PATH, 'default').doModal()
+    sys.modules.clear()
```

Comment:
Leftover?

Reply:
A lot of screensavers are doing this, so I decided to include it as well.
I have had various issues with the implementation and this was one of the measures that made things a lot more reliable.

Reply:
After some more isolated testing, I found two issues and removed the `sys.modules.clear()` call.
dagwieers/screensaver.turnoff#13
dagwieers/screensaver.turnoff#14

I hope they weren't needed for Krypton :-/